### DISCUSSION TITLE
Sticky-keys with tap-hold

### DISCUSSION THREAD

**Discussion by nicknimchuk** [kmonad/kmonad#570] [Q&A]
Original post, see first reply with actual confusion:

If I create a sticky key (say, left shift), then put it in a layer that I activate with "layer-toggle" on a different key, it does not appear to be active if I release the layer-toggle key. I'm trying to put sticky modifier keys into a layer so that I can access them with a quick key-combo but avoid having to keep them held down. That way, I can use the modifiers on keys that are masked by the new layer.

Is there a good way to accomplish this? I tried a few variations of sticky keys, around-next, and maybe others, but nothing seemed to work like I expected.

**Comment by nicknimchuk** [comment]
I figured out that my problem had nothing to do with layers, and everything to do with tap-hold (where I'm trying to emulate the QMK "auto-shift"). When using a sticky shift with a tap-hold button, the shift gets canceled before the tap event is processed. With the following configuration bits in `defalias`:
```
as_w (tap-hold 144 w S-w)
slsft (sticky-key 10001 lsft)
```
With `as_w` assigned to the `w` key and `slsft` to `ralt`, tapping the sticky shift followed by tapping the `w` key results in a lower-case `w` and generates the following debug log:
```
--------------------------------------------------------------------------------
Received event: Press <ralt>
Running hooks
Registering untimed hook: 13
Registering untimed hook: 14

--------------------------------------------------------------------------------
Received event: Release <ralt>
Running hooks
Registering 10001ms hook: 15

--------------------------------------------------------------------------------
Received event: Press <w>
Running hooks
Injecting event: Press <w>
Emitting: Press <sft>

--------------------------------------------------------------------------------
Rerunning event: Press <w>
Registering 3ms hook: 16
Running hooks
Block level set to: 1
Registering 2ms hook: 17
Registering untimed hook: 19
Registering 144ms hook: 18
Cancelling hook: 17
Tried cancelling expired hook: 16
Emitting: Release <sft>

--------------------------------------------------------------------------------
Received event: Release <w>
Running hooks
Emitting: Press <w>
Unblocking input stream, no stored events
Emitting: Release <w>
Tried cancelling expired hook: 18
Tried cancelling expired hook: 9
Tried cancelling expired hook: 15

--------------------------------------------------------------------------------
```

It appears that the sticky shift is cancelled by the tap-hold starting, but the event for that isn't generated until I release it. Is there a "right" way to handle this?

**Reply by markstos** [reply]
I use this feature in my QMK layout I was translating to KMonad. In my case, I have both "Control" and "Super" on my left thumb, with Control as a "one-shot" key and "Super" as  "Tap-hold" type of key.
So the way I type "Control-Super-Space" is to tap "Control" and then hold Super and type space. 

I realized in that case I can workaround the ability chain them by pressing the adjacent keys at the same time with my thumb.

**Comment by david-janssen** [comment]
Ahhhhh, it's always a bit difficult to really figure out what's going on internally from these printouts, but looking at it, it seems that `sticky-keys` don't really play nice with non-simple buttons. I think the `sticky-key` is wrapping around the computation of the taphold, but then terminates its action before the resolved taphold does its tap. This looks like a difficult fix.

**Reply by nicknimchuk** [reply]
It seems like it might even be a race condition. While playing around, I thought I had a solution with `around-next-timeout` where it worked as expected, but the "fix" stopped working after I changed some other keys and then reloaded kmonad. So yes, sounds like a difficult fix to me as well!

**Reply by markstos** [reply]
Since this seems like a bug, should this be converted into an issue? This combination is something QMK supports.

**Reply by knurd75gh** [reply]
Hello! I like kmonad a lot. Thanks for all the work.
I also have this very issue. I use a home row mod, kind of setup and can not use sticky-key shift for the home row keys.
